oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/tree/impl/AbstractTree.java

```diff
@@ -150,34 +150,63 @@ protected Iterable<String> getChildNames() {
 
     @Override
     public String toString() {
-        return toString(5);
+        return toJsonString(2);
     }
 
-    private String toString(int childNameCountLimit) {
-        StringBuilder sb = new StringBuilder();
-        sb.append(getPath()).append(": ");
-
-        sb.append('{');
-        for (PropertyState p : getProperties()) {
-            sb.append(' ').append(p).append(',');
+    /**
+     * Represents an OAK tree as a JSON object string
+     * When the representation is limited in depth, {"...":"..."} represents the limit.
+     * Note that, try-catch sentences are used to add "ERROR: "+e.getMessage() values in case of representation error.
+     * Note that, this is a recursive method.
+     * @param depth tree depth to represent. Use -1 for an unlimited depth.
+     * @return json object representation of the tree as a string
+     */
+    public String toJsonString(int depth) {
+        if (depth == 0) {
+            return quote("...");
         }
-
-        Iterator<String> names = this.getChildNames().iterator();
-        int count = 0;
-        while (names.hasNext() && ++count <= childNameCountLimit) {
-            sb.append(' ').append(names.next()).append(" = { ... },");
+        StringBuilder str = new StringBuilder();
+        str.append("{");
+        str.append(quote("_properties_") + ":{ ");
+        for (PropertyState ps : this.getProperties()) {
+            str.append(quote(ps.getName()) + ":");
+
+            if (ps.getType().isArray()) {
+                str.append("[ ");
+                for (int i = 0; i < ps.count(); i++) {
+                    try {
+                        str.append(quote(escapeJsonString(ps.getValue(Type.STRING, i))) + ",");
+                    } catch (Exception e) {
+                        str.append(quote("ERROR:" + escapeJsonString(e.getMessage())) + ",");
+                    }
+                }
+                str.deleteCharAt(str.length() - 1); //removing the space or the ,
+                str.append("],");
+            } else {
+                try {
+                    str.append(quote(escapeJsonString(ps.getValue(Type.STRING))) + ",");
+                } catch (Exception e) {
+                    str.append(quote("ERROR:" + escapeJsonString(e.getMessage())) + ",");
+                }
+            }
         }
-
-        if (names.hasNext()) {
-            sb.append(" ...");
+        str.deleteCharAt(str.length() - 1);  //removing the space or the ,
+        str.append("},");
+        for (Tree child : this.getChildren()) {
+            str.append(quote(child.getName()) + ":");
```

Comment:
I think it needs to be:

    str.append(quote(escapeJsonString(child.getName())) + ":");

Reply:
Applied with JsopBuilder.encode()